### PR TITLE
Migrate Layout File to Typescript

### PR DESCRIPTION
**Pull request by ong6**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Part of #1913 

**Overview of changes:**
Migrate Layouts file and Layouts Manager to typescript

**Anything you'd like to highlight/discuss:**
N/A

**Testing instructions:**
N/A

**Proposed commit message: (wrap lines at 72 characters)**
(rebase commit)

---

**Checklist:** :ballot_box_with_check:

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes

### CODE REVIEW

File: packages/core/src/Layout/Layout.ts [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+  config: any;
+
+  includedFiles: Set<string>;
+  layoutProcessed: string;
+  layoutPageBodyVariable: string;
+  layoutPageNavUuid: string;
+  headTop: any[];
+  headBottom: any[];
+  scriptBottom: any[];
+  layoutUserScriptsAndStyles: any[];
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:22]: Could we also avoid the `any[]` types here?

File: packages/core/src/Layout/Layout.ts [hunk 2]
```diff
@@ -41,21 +54,49 @@ class Layout {
...
+      };
 
-      const { variableProcessor, pluginManager, siteLinkManager } = this.config;
+      const {
+        variableProcessor,
+        pluginManager,
+        siteLinkManager,
+      }: {
+        variableProcessor: VariableProcessor;
+        pluginManager: any;
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:67]: We can use the defined `PluginManager` type here - It's exported from `PluginManager.ts`

File: packages/core/src/Layout/Layout.ts [hunk 3]
```diff
@@ -41,21 +54,49 @@ class Layout {
...
-        this.sourceFilePath, pageSources, layoutVariables);
-      nodeProcessor = new NodeProcessor(fileConfig, pageSources, variableProcessor,
-                                        pluginManager, siteLinkManager, this.layoutUserScriptsAndStyles,
-                                        'layout');
+        this.sourceFilePath,
+        pageSources,
+        layoutVariables,
+      );
+      nodeProcessor = new NodeProcessor(
+        fileConfig,
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:77]: Minor nit, why are all the params on a line on its own? 😅 Is there a reason for making this change throughout the file? ESLint doesn't seem to mind the original way

**Review by jovyntls — CHANGES_REQUESTED** [review]
Some comments on the types! Also, why are all the method call params (and anything comma-separated) split into a line of their own? Seems like an unnecessary change 😅

File: packages/core/src/Layout/Layout.ts (hunk 3 above)
Comment (ong6) [reply packages/core/src/Layout/Layout.ts:77]: i think my vs code was giving me some linting tips haha, so I went and made the change

File: .eslintignore [hunk 4]
```diff
@@ -41,6 +41,10 @@ packages/core/src/html/MarkdownProcessor.js
 packages/core/src/html/warnings.js
 packages/core/src/plugins/Plugin.js
 packages/core/src/plugins/PluginManager.js
+packages/core/src/Layout/index.js
+packages/core/src/Layout/Layout.js
+packages/core/src/Layout/LayoutManager.js
+
```
Comment (raysonkoh) [inline .eslintignore:47]: Nit: unnecessary space here

File: packages/core/src/Layout/Layout.ts (hunk 1 above)
Comment (raysonkoh) [reply packages/core/src/Layout/Layout.ts:22]: Just curious, what's the conclusion for this?

File: packages/core/src/Layout/Layout.ts [hunk 5]
```diff
@@ -98,6 +146,4 @@ class Layout {
   }
 }
 
-module.exports = {
-  Layout,
-};
+export { Layout };
```
Comment (raysonkoh) [inline packages/core/src/Layout/Layout.ts:149]: Possible to do `export class Layout {...}` instead?

File: packages/core/src/Layout/Layout.ts (hunk 3 above)
Comment (raysonkoh) [reply packages/core/src/Layout/Layout.ts:77]: I think it would be good to revert some of these formatting changes for better code readability.

File: packages/core/src/Layout/Layout.ts [hunk 6]
```diff
@@ -1,14 +1,38 @@
...
 
 const LAYOUT_PAGE_BODY_VARIABLE = 'content';
 
-class Layout {
-  constructor(sourceFilePath, config) {
+export type LayoutConfig = {
+  pluginManager: PluginManager;
+  siteLinkManager: SiteLinkManager;
+  variableProcessor: VariableProcessor;
+  externalManager: any;
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:17]:
```suggestion
  externalManager: ExternalManager;
```
Would it be better to use the `ExternalManager` class here?

File: packages/core/src/Layout/Layout.ts [hunk 7]
```diff
@@ -1,14 +1,38 @@
...
+  config: LayoutConfig;
+
+  includedFiles: Set<string>;
+  layoutProcessed: string;
+  layoutPageBodyVariable: string;
+  layoutPageNavUuid: string;
+  headTop: string[];
+  headBottom: string[];
+  scriptBottom: string[];
+  layoutUserScriptsAndStyles: string[] | undefined;
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:31]:
```suggestion
  layoutUserScriptsAndStyles?: string[];
```

File: packages/core/src/Layout/Layout.ts [hunk 8]
```diff
@@ -41,18 +63,24 @@ class Layout {
...
       pageSources = new PageSources();
       this.layoutPageBodyVariable = `{{${uuidv4()}-${uuidv4()}}}`;
-      const layoutVariables = { [LAYOUT_PAGE_BODY_VARIABLE]: this.layoutPageBodyVariable };
+      const layoutVariables = {
+        [LAYOUT_PAGE_BODY_VARIABLE]: this.layoutPageBodyVariable,
+      };
 
-      const { variableProcessor, pluginManager, siteLinkManager } = this.config;
+      const { variableProcessor, pluginManager, siteLinkManager }: {
+        variableProcessor: VariableProcessor, pluginManager: PluginManager, siteLinkManager: SiteLinkManager
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:71]: Is this line necessary? These types might be implicit since we are destructuring from `this.config` which defines the attributes as having these types. Correct me if I'm wrong 😅

File: packages/core/src/Layout/Layout.ts [hunk 9]
```diff
@@ -41,18 +63,24 @@ class Layout {
...
+      } = this.config;
 
       const nunjucksProcessed = variableProcessor.renderWithSiteVariables(
         this.sourceFilePath, pageSources, layoutVariables);
       nodeProcessor = new NodeProcessor(fileConfig, pageSources, variableProcessor,
                                         pluginManager, siteLinkManager, this.layoutUserScriptsAndStyles,
                                         'layout');
+
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:79]: Unnecessary whitespace change?

File: packages/core/src/Layout/Layout.ts [hunk 10]
```diff
@@ -65,8 +93,9 @@ class Layout {
     } while (triesLeft > 0 && numBodyVars > 1);
 
     if (triesLeft === 0) {
-      logger.error(
-        `Layout ${this.sourceFilePath} uses more than one {{ ${LAYOUT_PAGE_BODY_VARIABLE} }} variable.`);
+      error(
+        `Layout ${this.sourceFilePath} uses more than one {{ ${LAYOUT_PAGE_BODY_VARIABLE} }} variable.`,
+      );
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:98]: Unnecessary whitespace change? (Trailing comma and newline)

File: packages/core/src/Layout/Layout.ts [hunk 11]
```diff
@@ -75,11 +104,13 @@ class Layout {
     this.scriptBottom = nodeProcessor.scriptBottom;
 
     pageSources.addAllToSet(this.includedFiles);
-    await this.config.externalManager.generateDependencies(pageSources.getDynamicIncludeSrc(),
-                                                           this.includedFiles);
+    await this.config.externalManager.generateDependencies(
+      pageSources.getDynamicIncludeSrc(),
+      this.includedFiles,
+    );
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:110]: Is this also an unnecessary whitespace change?

File: packages/core/src/Layout/Layout.ts [hunk 12]
```diff
@@ -41,18 +63,24 @@ class Layout {
       };
       pageSources = new PageSources();
       this.layoutPageBodyVariable = `{{${uuidv4()}-${uuidv4()}}}`;
-      const layoutVariables = { [LAYOUT_PAGE_BODY_VARIABLE]: this.layoutPageBodyVariable };
+      const layoutVariables = {
+        [LAYOUT_PAGE_BODY_VARIABLE]: this.layoutPageBodyVariable,
+      };
```
Comment (jovyntls) [inline packages/core/src/Layout/Layout.ts:68]: Unnecessary whitespace change

**Review by jovyntls — CHANGES_REQUESTED** [review]
Thanks for the PR @ong6! Can we avoid unnecessary whitespace changes (e.g. putting all variables in a newline?

File: packages/core/src/Layout/Layout.ts (hunk 11 above)
Comment (ong6) [reply packages/core/src/Layout/Layout.ts:110]: this exceeds the max line length, so i think this one is ok

Comment (jovyntls) [reply packages/core/src/Layout/Layout.ts:104]: iirc our max line length is 110? The existing one is fine by my linter

Comment (ong6) [reply packages/core/src/Layout/Layout.ts]: ok yeah, i've reverted it to the original now

File: packages/core/src/Layout/LayoutManager.ts [hunk 13]
```diff
@@ -81,6 +85,6 @@ class LayoutManager {
   }
 }
 
-module.exports = {
+export {
   LayoutManager,
 };
```
Comment (jovyntls) [inline packages/core/src/Layout/LayoutManager.ts:90]: Nit: the export could be `export class LayoutManager` to keep consistency with Layout

**Review by jovyntls — APPROVED** [review]
LGTM!

**Review by jovyntls — COMMENTED** [review]
@ong6 Sorry I missed this earlier - could you update the imports of the files you migrated to use ES6 imports instead of `require`?

File: .eslintignore [hunk 14]
```diff
@@ -57,6 +57,13 @@ packages/core/src/plugins/default/markbind-plugin-anchors.js
 packages/core/src/plugins/default/markbind-plugin-plantuml.js
 packages/core/src/plugins/default/markbind-plugin-shorthandSyntax.js
 packages/core/src/plugins/default/markbind-plugin-tree.js
+packages/core/src/Layout/index.js
+packages/core/src/Layout/Layout.js
+packages/core/src/Layout/LayoutManager.js
+packages/core/src/plugins/default/markbind-plugin-anchors.js
+packages/core/src/plugins/default/markbind-plugin-plantuml.js
+packages/core/src/plugins/default/markbind-plugin-shorthandSyntax.js
+packages/core/src/plugins/default/markbind-plugin-tree.js
```
Comment (jovyntls) [inline .eslintignore:66]: This seems like a rebase error @ong6 - can we remove the duplicates here?

**Review by jovyntls — CHANGES_REQUESTED** [review]
@ong6 Please ensure that you've fixed the merge conflicts

**Comment by tlylt** [comment]
Almost there now, let's try to get this one merged soon

**Review by jovyntls — CHANGES_REQUESTED** [review]
Just a few more @ong6 !

File: packages/core/src/Layout/LayoutManager.ts [hunk 15]
```diff
@@ -1,14 +1,18 @@
...
 
-const logger = require('../utils/logger');
+import * as logger from '../utils/logger';
 
 const FRONTMATTER_NONE_ATTR = 'none';
 
-class LayoutManager {
-  constructor(config) {
+export class LayoutManager {
+  layouts: { [key: string]: Layout } = {};
```
Comment (jovyntls) [inline packages/core/src/Layout/LayoutManager.ts:11]:
```suggestion
  layouts: Record<string, Layout>;
```
For consistency

File: packages/core/src/Layout/LayoutManager.ts [hunk 16]
```diff
@@ -69,18 +73,14 @@ class LayoutManager {
       return pageContent;
     }
 
-    return this.layouts[name].insertPage(pageContent, pageNav, pageIncludedFiles);
+    return this.layouts[name].insertPage(pageContent, pageNav, new Set<string>(pageIncludedFiles));
```
Comment (jovyntls) [inline packages/core/src/Layout/LayoutManager.ts:76]: Why do we create a new instance of `pageIncludedFiles` here? Is it to avoid passing by reference?

File: packages/core/src/Page/index.ts [hunk 17]
```diff
@@ -443,7 +440,8 @@ export class Page {
    */
   buildPageNav(content: string) {
     const isFmPageNavSpecifierValid = this.isPageNavigationSpecifierValid();
-    const doesLayoutHavePageNav = this.pageConfig.layoutManager.layoutHasPageNav(this.layout);
+    const doesLayoutHavePageNav = this.layout ? this.pageConfig.layoutManager.layoutHasPageNav(
+      this.layout) : false;
```
Comment (jovyntls) [inline packages/core/src/Page/index.ts:444]: I think we can avoid all of these typing issues with `this.layout`, and the additions of `?? LAYOUT_DEFAULT_NAME` below, by asserting that the `layout` attribute exists with a `!` operator. 
You'll need to modify [this](https://github.com/ong6/markbind/blob/f11595a93c3cd3dea8f815606875a9498461062a/packages/core/src/Page/PageConfig.ts#L79) too (just delete the question mark)

Feel free to trace to verify that this is safe!

**Review by jovyntls — CHANGES_REQUESTED** [review]
Hi @ong6, I think you need to fix the rebase - the tests are failing because your .gitignore/.eslintignore are not ignoring the files that they should, probably because of an error when you rebased

**Comment by ong6** [comment]
Will create new PR for this
